### DISCUSSION TITLE
Accessing Row items by name

### DISCUSSION THREAD

**Discussion by lelit** [sqlalchemy/sqlalchemy#10756] [Usage Questions] [answered]
I have a library that I use to expose arbitrary SA queries to web services: basically, given a _generic_ query it applies filters and sorters from a web request's params, and then returns a list of dictionaries.

Recently, to support new style SA 2 selects, that may be _composed_ by mapped entities and standalone columns such as `select(MyEntity, MyOtherEntity.some_column`, I did use the private attribute `Row._mapping` (see [here](https://gitlab.com/metapensiero/metapensiero.sqlalchemy.proxy/-/blob/master/src/metapensiero/sqlalchemy/proxy/core.py#L138-148) for details) to operate on each row item trasforming it into a "flat" list of values.

Today I found that that (dirty?) trick falls short when the query contains an `aliased` entity: in that case the `_mapping` attribute is an empty dictionary, so my code does not know how to proceed.

Here is a demo of the problem:

```python
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy import create_engine
from sqlalchemy import orm
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import sessionmaker


metadata = MetaData()


persons = Table(
    'persons', metadata,
    Column('id', Integer, primary_key=True),
    Column('firstname', String),
    Column('lastname', String),
)


class Person:
    def __init__(self, firstname, lastname):
        self.firstname = firstname
        self.lastname = lastname


orm.registry().map_imperatively(Person, persons)


engine = create_engine('sqlite:///:memory:')
session = sessionmaker(autoflush=False, autocommit=False, bind=engine)

metadata.create_all(engine)

with session() as sas:
    me = Person('Lele', 'Gaifax')
    sas.add(me)
    sas.commit()

with session() as sas:
    query = select(Person)
    result = sas.execute(query)
    for row in result:
        print('row._mapping is', row._mapping)
        print('row is', row)

with session() as sas:
    alias = orm.aliased(Person)
    query = select(alias)
    result = sas.execute(query)
    for row in result:
        print('row._mapping is', row._mapping)
        print('row is', row)
```

which prints

```
row._mapping is {'Person': <__main__.Person object at 0x7ffff5ad7a60>}
row is (<__main__.Person object at 0x7ffff5ad7a60>,)
row._mapping is {}
row is (<__main__.Person object at 0x7ffff5ad7700>,)
```

The referenced code in my library wants to transform say `select(Person, OtherEntity.some_column)` into a list of dictionaries like
```python
[
  { 'firstname': 'Lele', 'lastname': 'Gaifax', 'some_column': 42 }
]
```

I obviously understand that the `_mapping` is _private_ for a reason and so I may be abusing it, but is there a better alternative?

**Comment by CaselIT** [comment]
Hi,

`_mapping` is not private, inspired by other attributes that are on named tuple like objects, such as `_fields` etc. It's documented here https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row

This looks like a bug. What does _fields returns on row?

**Reply by CaselIT** [reply]
Or `row._mapping.keys()`. The issue may be that the key is `None` and somewhere there is a `if k is not None`

**Reply by zzzeek** [reply]
the anonymous alias has no defined name for the purposes of the row

**Comment by zzzeek** [comment, marked as answer]
you need to give your alias a name, then it will be in the mapping:

```

palias = aliased(Person, name="p1")

```

**Reply by lelit** [reply]
Thanks Michael, that made it :heart: !
